Sources/SwiftLanguageService/CodeActions/RemoveUnusedImports.swift

```diff
@@ -0,0 +1,278 @@
+
+        if try await temporaryDocumentHasErrorDiagnostic() {
+          // The file now has syntax error where it didn't before. Add the import decl back in again.
+          let addImportReq = sourcekitd.dictionary([
+            keys.name: temporaryDocUri.pseudoPath,
+            keys.enableSyntaxMap: 0,
+            keys.enableStructure: 0,
+            keys.enableDiagnostics: 0,
+            keys.syntacticOnly: 1,
+            keys.offset: startOffset,
+            keys.length: 0,
+            keys.sourceText: importDecl.description,
+          ])
+          _ = try await self.send(sourcekitdRequest: \.editorReplaceText, addImportReq, snapshot: nil)
+
+          continue
+        }
+
+        declsToRemove.append(importDecl)
+      }
+
+      guard let sourceKitLSPServer else {
+        throw ResponseError.unknown("Connection to the editor closed")
+      }
+
+      let edits = declsToRemove.reversed().map { importDecl in
+        var range = snapshot.range(of: importDecl)
+
+        let isAtStartOfFile = importDecl.previousToken(viewMode: .sourceAccurate) == nil
+
+        if isAtStartOfFile {
+          // If this is at the start of the source file, keep its leading trivia since we should consider those as a
+          // file header instead of belonging to the import decl.
+          range = snapshot.position(of: importDecl.positionAfterSkippingLeadingTrivia)..<range.upperBound
+        }
+
+        // If we are removing the first import statement in the file and it is followed by a newline (which will belong
+        // to the next token), remove that newline as well so we are not left with an empty line at the start of the
+        // source file.
+        if isAtStartOfFile,
+          let nextToken = importDecl.nextToken(viewMode: .sourceAccurate),
+          nextToken.leadingTrivia.first?.isNewline ?? false
+        {
+          let nextTokenWillBeRemoved =
+            nextToken.ancestorOrSelf(mapping: { (node) -> Syntax? in
+              guard let importDecl = node.as(ImportDeclSyntax.self), declsToRemove.contains(importDecl) else {
+                return nil
+              }
+              return node
+            }) != nil
+          if !nextTokenWillBeRemoved {
+            range = range.lowerBound..<snapshot.position(of: nextToken.position.advanced(by: 1))
+          }
+        }
+
+        return TextEdit(range: range, newText: "")
+      }
+      let applyResponse = try await sourceKitLSPServer.sendRequestToClient(
+        ApplyEditRequest(
+          edit: WorkspaceEdit(
+            changes: [snapshot.uri: edits]
+          )
+        )
+      )
+      if !applyResponse.applied {
+        let reason: String
+        if let failureReason = applyResponse.failureReason {
+          reason = " reason: \(failureReason)"
+        } else {
+          reason = ""
+        }
+        logger.error("client refused to apply edit for removing unused imports: \(reason)")
+      }
+    } cleanup: {
```

Comment:
Poor man's defer because of the async call I assume?

Reply:
Correct. Good reminder to add a `#if swift(>=6.4)` warning to remove this once we have https://github.com/swiftlang/swift-evolution/blob/main/proposals/0493-defer-async.md